crates/biome_js_analyze/src/lint/nursery/no_await_in_loop.rs

```diff
@@ -0,0 +1,144 @@
+        if let AnyLoopNode::JsForOfStatement(for_of) = loop_node {
+            if for_of.await_token().is_some() {
+                return None;
+            }
+        }
+
+        let mut preorder = loop_node.syntax().preorder();
+        while let Some(event) = preorder.next() {
+            match event {
+                WalkEvent::Enter(node) => {
+                    if is_boundary(&node) {
+                        preorder.skip_subtree();
+                    }
+
+                    // skip valid case: for await
+                    // e.g. `while (baz) { for await (x of xs)}`
+                    if let Some(for_of) = JsForOfStatement::cast(node.clone()) {
+                        if for_of.await_token().is_some() {
+                            return Some(node);
+                        }
+                    }
+
+                    // skip valid case: binding in `for`
+                    // e.g. `async function foo() { for (var i = await bar; i < n; i++) {  } }`
+                    if JsVariableDeclaration::can_cast(node.kind()) {
+                        if let Some(parent) = node.parent() {
+                            if JsForStatement::can_cast(parent.kind()) {
+                                preorder.skip_subtree();
+                            }
+                        }
+                    }
+
+                    // skip valid case: bidning in `for in`
+                    // `async function foo() { for (var bar = await baz in qux) {} }`
+                    if JsForVariableDeclaration::can_cast(node.kind()) {
+                        preorder.skip_subtree();
+                    }
+
+                    if JsAwaitExpression::can_cast(node.kind()) {
+                        // skip valid cases: expression in `for in` and `for of`
+                        // - `async function foo() { for (var bar in await baz) { } }`
+                        // - `async function foo() { for (var bar of await baz) { } }`
+                        if let Some(parent) = node.parent() {
+                            if JsForOfStatement::can_cast(parent.kind())
+                                || JsForInStatement::can_cast(parent.kind())
+                            {
+                                continue;
+                            }
+                        }
+                        return Some(node);
+                    }
+                }
+                WalkEvent::Leave(_) => {}
+            }
+        }
+        None
+    }
+
+    fn diagnostic(_ctx: &RuleContext<Self>, state: &Self::State) -> Option<RuleDiagnostic> {
+        Some(
+            RuleDiagnostic::new(
+                rule_category!(),
+                state.text_range(),
+                markup! { "Avoid using "<Emphasis>"await"</Emphasis>" inside loops." },
+            )
+            .note(markup! {
+                "Using "<Emphasis>"await"</Emphasis>" inside loops might cause performance issues or unintended sequential execution."
+            }),
```

Comment:
As for our rule pillars, this diagnostic misses the third pillar: tell the use what to do to solve the violation